UniversalAlgebra/Structures/Graphs.lagda

```diff
@@ -0,0 +1,199 @@
+  i (inr 𝑓) a x = goal
+   where
+   homop : h (op 𝑨 𝑓 (a ∘ inl)) ≡ op 𝑩 𝑓 (h ∘ (a ∘ inl))
+   homop = (snd hhom) 𝑓 (a ∘ inl)
+
+   goal : op 𝑩 𝑓 (h ∘ (a ∘ inl)) ≡ h (a (inr 𝟙.𝟎))
+   goal = op 𝑩 𝑓 (h ∘ (a ∘ inl)) ≡⟨ sym homop ⟩
+          h (op 𝑨 𝑓 (a ∘ inl))   ≡⟨ cong h x ⟩
+          h (a (inr 𝟙.𝟎))         ∎
+
+  ii : is-hom-op (Gr 𝑨) (Gr 𝑩) h
+  ii = λ ()
+
+
+ Grhom→hom : hom (Gr 𝑨) (Gr 𝑩) → hom 𝑨 𝑩
+ Grhom→hom (h , hhom) = h , (i , ii)
+  where
+  i : is-hom-rel 𝑨 𝑩 h
+  i R a x = fst hhom (inl R) a x
+  ii : is-hom-op 𝑨 𝑩 h
+  ii f a = goal
+   where
+   split : arity 𝐹 f ⊎ 𝟙 → carrier 𝑨
+   split (inl x) = a x
+   split (inr y) = op 𝑨 f a
+   goal : h (op 𝑨 f a) ≡ op 𝑩 f (λ x → h (a x))
+   goal = sym (fst hhom (inr f) split refl)
+
+
+
+{- Lemma III.1. Let S be a signature and A be an S-structure.
+Let Σ be a finite set of identities such that A ⊧ Σ. For every
+instance X of CSP(A), one can compute in polynomial time an
+instance Y of CSP(A) such that Y ⊧ Σ and | Hom(X , A)| = |Hom(Y , A)|. -}
+
+
+module _ {𝐹 : signature}{χ : Level} where
+
+ data Term (X : Type χ ) : Type χ  where
+  ℊ : X → Term X    -- (ℊ for "generator")
+  node : (f : symbol 𝐹)(𝑡 : (arity 𝐹) f → Term X) → Term X
+
+ open Term public
+
+\end{code}
+
+When we interpret a term in a structure we call the resulting function a *term operation*.  Given a term `p` and a structure `𝑨`, we denote by `𝑨 ⟦ p ⟧` the *interpretation* of `p` in `𝑨`.  This is defined inductively as follows.
+
+1. If `p` is a variable symbol `x : X` and if `a : X → ∣ 𝑨 ∣` is a tuple of elements of `∣ 𝑨 ∣`, then `𝑨 ⟦ p ⟧ a := a x`.
+
+2. If `p = 𝑓 𝑡`, where `𝑓 : ∣ 𝑆 ∣` is an operation symbol, if `𝑡 : ∥ 𝑆 ∥ 𝑓 → 𝑻 X` is a tuple of terms, and if `a : X → ∣ 𝑨 ∣` is a tuple from `𝑨`, then we define `𝑨 ⟦ p ⟧ a = 𝑨 ⟦ 𝑓 𝑡 ⟧ a := (𝑓 ̂ 𝑨) (λ i → 𝑨 ⟦ 𝑡 i ⟧ a)`.
+
+Thus the interpretation of a term is defined by induction on the structure of the term, and the definition is formally implemented as follows.
+
+\begin{code}
+
+module _ {𝐹 𝑅 : signature}{χ : Level}{X : Type χ} where
+
+ _⟦_⟧ : (𝑨 : structure 𝐹 {ℓ₀} 𝑅 {ℓ₀}) → Term X → (X → carrier 𝑨) → carrier 𝑨
+ 𝑨 ⟦ ℊ x ⟧ = λ η → η x
+ 𝑨 ⟦ node 𝑓 𝑡 ⟧ = λ η → ((op 𝑨) 𝑓) (λ i → (𝑨 ⟦ 𝑡 i ⟧) η)
+
+ _⊧_≈_ : structure 𝐹 𝑅 → Term X → Term X → Type _
+ 𝑨 ⊧ p ≈ q = 𝑨 ⟦ p ⟧ ≈ 𝑨 ⟦ q ⟧
+
+ _⊧_≋_ : Pred(structure 𝐹 𝑅) ℓ₀ → Term X → Term X → Type _
+ 𝒦 ⊧ p ≋ q = {𝑨 : structure 𝐹 𝑅} → 𝒦 𝑨 → 𝑨 ⊧ p ≈ q
+
+
+ Th : Pred (structure 𝐹 𝑅) ℓ₀ → Pred(Term X × Term X) (ℓ₁ ⊔ χ)
+ Th 𝒦 = λ (p , q) → 𝒦 ⊧ p ≋ q
+
+ Mod : Pred(Term X × Term X) (χ ⊔ ℓ₀) → Pred(structure 𝐹 {ℓ₀} 𝑅 {ℓ₀}) χ
+ Mod ℰ = λ 𝑨 → ∀ p q → (p , q) ∈ ℰ → 𝑨 ⊧ p ≈ q
+
+ fMod : {n : ℕ} → (Fin n → (Term X × Term X)) → Pred(structure 𝐹 {ℓ₀} 𝑅 {ℓ₀}) χ
+ fMod ℰ = λ 𝑨 → ∀ i → 𝑨 ⊧ fst (ℰ i) ≈ snd (ℰ i)
+
+\end{code}
+
+The entailment ℰ ⊢ p ≈ q is valid iff p ≈ q holds in all models that satify all equations in ℰ.
+
+\begin{code}
+
+module _ {𝐹 𝑅 : signature} where
+
+ record _⇛_⇚_ (𝑩 𝑨 𝑪 : structure 𝐹 𝑅) : Type ℓ₀ where
```

Comment:
This is equivalence specialized - why not re-use the stdlib definition? The notation can be kept the same, just the definition changed.

Reply:
I'm not sure what you mean by "equivalence specialized."  Are you sure this is something that exists in the std-lib?

As for the arrow notation I'm using, I actually don't like it... It was just the first thing I thought of, but we should find something better... Anyway, this is just a start of my effort to formalize some theorems from our CSP work.

Reply:
It's the old Function.Inverse https://agda.github.io/agda-stdlib/Function.Inverse.html.  The more modern version is Inverse in https://agda.github.io/agda-stdlib/Function.Bundles.html which does have a couple of extra fields, but those are usually 'free'.  So this is Hom B A <-> Hom C A .